benchmark/tangurnis/code/tangurnis.cc

```diff
@@ -581,43 +537,92 @@ namespace aspect
+      << material_model->parameter_a();
+
+    // pad the first line to the same number of columns as the data below to make MATLAB happy
+    for (unsigned int i=4; i<7+this->get_heating_model_manager().get_active_heating_models().size(); ++i)
+      f << " -1";
 
+    f << std::endl;
     f << std::scientific;
 
 
     const QGauss<dim> quadrature_formula (this->get_fe().base_element(this->introspection().base_elements.velocities).degree+2);
 
     const unsigned int n_q_points =  quadrature_formula.size();
     FEValues<dim> fe_values (this->get_mapping(), this->get_fe(),  quadrature_formula,
-                             update_JxW_values | update_values | update_quadrature_points);
+                             update_JxW_values | update_values    |
+                             update_gradients  | update_quadrature_points);
 
-    std::vector<Tensor<1, dim> > velocity_values (quadrature_formula.size());
-    std::vector<double>         temperature_values (quadrature_formula.size());
-    std::vector<double>         pressure_values (quadrature_formula.size());
+    MaterialModel::MaterialModelInputs<dim> in(fe_values.n_quadrature_points, this->n_compositional_fields());
+    MaterialModel::MaterialModelOutputs<dim> out(fe_values.n_quadrature_points, this->n_compositional_fields());
+
+    std::vector<std::vector<double> > composition_values (this->n_compositional_fields(),std::vector<double> (n_q_points));
+
+    const std::list<std_cxx11::shared_ptr<HeatingModel::Interface<dim> > > &heating_model_objects = this->get_heating_model_manager().get_active_heating_models();
+
+    std::vector<HeatingModel::HeatingModelOutputs> heating_model_outputs (heating_model_objects.size(),
+                                                                          HeatingModel::HeatingModelOutputs (n_q_points, this->n_compositional_fields()));
 
     typename DoFHandler<dim>::active_cell_iterator
     cell = this->get_dof_handler().begin_active(),
     endc = this->get_dof_handler().end();
     for (; cell != endc; ++cell)
       {
         fe_values.reinit (cell);
-        fe_values[this->introspection().extractors.velocities].get_function_values (this->get_solution(), velocity_values);
-        fe_values[this->introspection().extractors.pressure].get_function_values (this->get_solution(), pressure_values);
-        fe_values[this->introspection().extractors.temperature].get_function_values (this->get_solution(), temperature_values);
+        fe_values[this->introspection().extractors.temperature].get_function_values (this->get_solution(), in.temperature);
+        fe_values[this->introspection().extractors.pressure].get_function_values (this->get_solution(), in.pressure);
+        fe_values[this->introspection().extractors.velocities].get_function_values (this->get_solution(), in.velocity);
+        fe_values[this->introspection().extractors.pressure].get_function_gradients (this->get_solution(), in.pressure_gradient);
```

Comment:
We might not even need it in this specific case, but the pressure gradient is needed for example in the version of the adiabatic heating that is used by default if the user does not choose to use the simplified adiabatic heating. Hence, everywhere where we compute heating model outputs we normally also fill the pressure gradients. 
We can think about changing this in the future, but it is used in many other places in the code and probably something for a different pull request.